File: house_robber.rs
struct Solution;

impl Solution {
    pub fn rob(nums: Vec<i32>) -> i32 {
        let (mut prev, mut prev2) = (0,0);
        for n in nums {
            let curr = std::cmp::max(n + prev2, prev);
            prev2 = prev;
            prev = curr;
        }
        prev
    }
}

fn main() {
    let nums = vec![1,2,3,1];
    print!("{}", Solution::rob(nums)); //expect 4
}

File: other/house_robber.cpp
#include "leetcode.h"

class Solution {
public:
  int rob(vector<int> &nums) {
    // space-optimized dynamic programming approach
    int prev = 0, prev2 = 0, curr = 0;
    for (auto i : nums) {
      curr = max(prev, i + prev2);
      prev2 = prev;
      prev = curr;
    }
    return curr;
  }
};

int main() {
  Solution obj;
  vector<int> nums = {1, 2, 3, 1};
  cout << obj.rob(nums); // expect 4
}
